### PR TITLE
feat(RELEASE-1990): add pulp-push-disk-images python script

### PR DESCRIPTION
**Pull request by happybhati**
This commit adds a python script to replicate the functionality of the inline bash script in the pulp-push-disk-images internal task in the catalog repo. The task will be updated to use this python module instead.

It also adds a subprocess_cmd helper module for running external commands with optional stderr logging.

Assisted-By: Cursor

### CODE REVIEW

File: scripts/python/tasks/internal/pulp_push_disk_images.py [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+) -> None:
+    """Pull one OCI artifact and stage mapped files for Pulp upload."""
+    pull_spec = require_json_field(component, "containerImage")
+    destination_name = require_json_field(component, "staged", "destination")
+    destination = disk_image_dir / str(destination_name) / "FILES"
+    destination.mkdir(parents=True, exist_ok=True)
+
+    with tempfile.TemporaryDirectory() as download_dir:
+        download = Path(download_dir)
+        auth_file = Path(tempfile.mkstemp()[1])
```
Comment (swickersh) [inline scripts/python/tasks/internal/pulp_push_disk_images.py]:
```suggestion
        auth_file = file.make_tempfile_path("oras-auth-")
```
Maybe use the one from file module

File: scripts/python/tasks/internal/pulp_push_disk_images.py [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+            "Needed to publish to customer portal"
+        )
+        print(msg, file=sys.stderr)
+        with stderr_path.open("a", encoding="utf-8") as errf:
+            errf.write(msg + "\n")
+        raise _StepFailure("exit 1")
+
+    with tempfile.TemporaryDirectory() as disk_dir_name:
+        disk_image_dir = Path(disk_dir_name)
+        os.environ["DISK_IMAGE_DIR"] = str(disk_image_dir)
```
Comment (swickersh) [inline scripts/python/tasks/internal/pulp_push_disk_images.py]: It doesn't seem like this is ever used anywhere. The disk_image_dir variable is used, but I don't see anything reading the value from the env.

File: scripts/python/tasks/internal/pulp_push_disk_images.py [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+        auth_file = Path(tempfile.mkstemp()[1])
+        try:
+            auth_out = run_cmd(
+                ["select-oci-auth", str(pull_spec)],
+                check=True,
+            ).stdout
+            auth_file.write_text(auth_out, encoding="utf-8")
+            run_cmd(
+                [
+                    "oras",
```
Comment (swickersh) [inline scripts/python/tasks/internal/pulp_push_disk_images.py]: There is an oras helper that this might be a nice addition to

File: scripts/python/tasks/internal/pulp_push_disk_images.py [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+        components[0].get("staged", {}).get("version")
+        if isinstance(components[0], dict)
+        else None
+    )
+    if not version:
+        msg = (
+            "Error: version not specified in .components[0].staged.version. "
+            "Needed to publish to customer portal"
+        )
+        print(msg, file=sys.stderr)
```
Comment (swickersh) [inline scripts/python/tasks/internal/pulp_push_disk_images.py]: This print seems redundant

File: scripts/python/tasks/internal/pulp_push_disk_images.py [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+
+    for path, content in (
+        (exodus_gw_cert, exodus_cert),
+        (exodus_gw_key, exodus_key),
+        (pulp_cert_file, pulp_cert),
+        (pulp_key_file, pulp_key),
+        (udcache_cert, udc_cert),
+        (udcache_key, udc_key),
+    ):
+        path.write_text(content, encoding="utf-8")
```
Comment (swickersh) [inline scripts/python/tasks/internal/pulp_push_disk_images.py]: didn't these all already get written above with the make_tempfile_path?

Comment (happybhati) [reply scripts/python/tasks/internal/pulp_push_disk_images.py]: Removed it. Thanks for spotting that.

File: scripts/python/tasks/internal/pulp_push_disk_images.py (hunk 4 above)
Comment (happybhati) [reply scripts/python/tasks/internal/pulp_push_disk_images.py]: Dropped the extra print. Thanks.

File: scripts/python/tasks/internal/pulp_push_disk_images.py (hunk 3 above)
Comment (happybhati) [reply scripts/python/tasks/internal/pulp_push_disk_images.py]: Moved the pull logic into oras_pull() in oras_utils.py so it matches the rest of the OCI helpers. Appreciate the nudge — much cleaner.

File: scripts/python/tasks/internal/pulp_push_disk_images.py (hunk 2 above)
Comment (happybhati) [reply scripts/python/tasks/internal/pulp_push_disk_images.py]: You're right, nothing reads it in Python — we pass disk_image_dir directly. Removed the export; leftover from the bash port. Thanks.

File: scripts/python/tasks/internal/pulp_push_disk_images.py (hunk 1 above)
Comment (happybhati) [reply scripts/python/tasks/internal/pulp_push_disk_images.py]: Switched to file.make_tempfile_path inside oras_pull(). Thanks for pointing me at the existing helper.

File: scripts/python/helpers/oras_utils.py [hunk 6]
```diff
@@ -23,6 +29,37 @@ def oras_login(registry: str, username: str, password: str) -> None:
     )
 
 
+def oras_pull(
```
Comment (swickersh) [inline scripts/python/helpers/oras_utils.py:57]: Somehow oras_utils.py landed without unit tests prior to your change. (it was me :shamehide)
Could you please add a test file for this function? I can add tests for the other functions later.

Comment (happybhati) [reply scripts/python/helpers/oras_utils.py:57]: Yes for sure, I am on it. Thanks for pointing 🙂

File: scripts/python/tasks/internal/pulp_push_disk_images.py [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+                str(pulp_key_file),
+                "--udcache-url",
+                udc_url,
+            ],
+            env=env,
+            stderr_path=stderr_path,
+            check=True,
+        )
+
+        component_destinations = [
```
Comment (swickersh) [inline scripts/python/tasks/internal/pulp_push_disk_images.py]: This looks pre-existing from bash, but worth fixing here while we're rewriting. `component_destinations` is a flat list of all staged files indexed by component position — if any component has more than one staged file, subsequent components will receive the wrong content directory passed to `developer_portal_wrapper`.

Since the component already carries its destination in `staged.destination`, you can derive the directory directly and drop the `component_destinations` list entirely:

```python
for component in components:
    if not isinstance(component, dict):
        continue
    destination_name = require_json_field(component, "staged", "destination")
    content_directory = disk_image_dir / str(destination_name) / "FILES"
    process_component_for_developer_portal(
        component,
        content_directory,
        cgw_hostname,
        env=env,
        stderr_path=stderr_path,
        run_cmd=run_cmd,
    )
```

Also worth adding a unit test for the multi-component edge case (two components, first one with two staged files) to make sure the directories don't get crossed.

Cursor on behalf of Scott

Comment (swickersh) [reply scripts/python/tasks/internal/pulp_push_disk_images.py]: Cursor highlighted this one while reviewing with me. If it's truly pre-existing we could create a Jira and tackle it later.

File: scripts/python/helpers/oras_utils.py (hunk 6 above)
Comment (happybhati) [reply scripts/python/helpers/oras_utils.py:57]: Added `test_oras_utils.py` with tests for the auth/oras flow and temp file cleanup.

File: scripts/python/tasks/internal/pulp_push_disk_images.py (hunk 7 above)
Comment (happybhati) [reply scripts/python/tasks/internal/pulp_push_disk_images.py]: This was pre-existing in the bash task — the Python port carried over the same `component_destinations[$i]` indexing. Fixed here to use `staged.destination` per component, and added a unit test (two components, first with two staged files). Thanks for catching it.

**Review by swickersh — DISMISSED** [review]

File: scripts/python/helpers/oras_utils.py [hunk 8]
```diff
@@ -23,6 +29,37 @@ def oras_login(registry: str, username: str, password: str) -> None:
...
+    pull_spec: str,
+    download_dir: Path,
+    *,
+    run_cmd: RunCmd = subprocess_cmd.run_cmd,
```
Comment (johnbieren) [inline scripts/python/helpers/oras_utils.py]: Martin made a good point on one of my PRs about style like this - we never really are going to use it with a different function. If you ask cursor, it can adjust and have (in this case) `subprocess_cmd.run_cmd` be hardcoded and still have passing tests

Comment (happybhati) [reply scripts/python/helpers/oras_utils.py]: Thanks for the pointer.

Removed the `run_cmd` parameter from `oras_pull` and hardcoded `subprocess_cmd.run_cmd`. Updated tests to monkeypatch instead.

**Review by swickersh — DISMISSED** [review]

File: scripts/python/helpers/oras_utils.py [hunk 9]
```diff
@@ -23,6 +26,36 @@ def oras_login(registry: str, username: str, password: str) -> None:
...
+                "pull",
+                "--registry-config",
+                str(auth_file),
+                str(pull_spec),
+            ],
+            cwd=download_dir,
+            stderr_path=stderr_path,
+            check=True,
+        )
+    finally:
```
Comment (johnbieren) [inline scripts/python/helpers/oras_utils.py:83]: so won't this never fail? Is that intended? I would think we would want the function to call if the subprocess fails

File: scripts/python/tasks/internal/pulp_push_disk_images.py [hunk 10]
```diff
@@ -0,0 +1,455 @@
...
+DEFAULT_EXODUS_MOUNT = Path("/mnt/exodusGwSecret")
+DEFAULT_PULP_MOUNT = Path("/mnt/pulpSecret")
+DEFAULT_UDCACHE_MOUNT = Path("/mnt/udcacheSecret")
+DEFAULT_WORKLOADS_MOUNT = Path("/mnt/redhat-workloads-token")
+DEFAULT_CGW_MOUNT = Path("/mnt/cgwSecret")
+
+RunCmd = Callable[..., subprocess.CompletedProcess[str]]
+
+
+class _StepFailure(ValueError):
```
Comment (johnbieren) [inline scripts/python/tasks/internal/pulp_push_disk_images.py]: This sounds like it is trying to achieve the same thing that `CheckStepError` in the tekton.py helper already does

File: scripts/python/tasks/internal/pulp_push_disk_images.py [hunk 11]
```diff
@@ -0,0 +1,455 @@
...
+    """Raised when a step fails the same way as bash ``set -e`` + ``exitfunc``."""
+
+    def __init__(self, command: str, *, exit_code: int = 1, line: int = 0) -> None:
+        self.command = command
+        self.exit_code = exit_code
+        self.line = line
+        super().__init__(command)
+
+
+def read_mounted_text(mount: Path, filename: str) -> str:
```
Comment (johnbieren) [inline scripts/python/tasks/internal/pulp_push_disk_images.py]: Please import and use https://github.com/konflux-ci/release-service-utils/blob/main/scripts/python/helpers/authentication.py#L25 instead

File: scripts/python/tasks/internal/pulp_push_disk_images.py [hunk 12]
```diff
@@ -0,0 +1,455 @@
...
+    """Return the stripped contents of *filename* under *mount*."""
+    return (mount / filename).read_text(encoding="utf-8").strip()
+
+
+def check_certificates(
```
Comment (johnbieren) [inline scripts/python/tasks/internal/pulp_push_disk_images.py]: Why not use https://github.com/konflux-ci/release-service-utils/blob/main/scripts/python/helpers/push_artifacts.py#L81 ?

File: scripts/python/helpers/oras_utils.py (hunk 9 above)
Comment (happybhati) [reply scripts/python/helpers/oras_utils.py:83]: Good catch on double-checking this — subprocess failures do propagate (`check=True`). Added a test to make that explicit and a brief comment on the `finally` block. Thanks!

File: scripts/python/tasks/internal/pulp_push_disk_images.py (hunk 10 above)
Comment (happybhati) [reply scripts/python/tasks/internal/pulp_push_disk_images.py]: Agreed — removed `_StepFailure` and switched to `CheckStepError` + `write_failure_result`. Thanks for pointing at the existing tekton helper!

File: scripts/python/tasks/internal/pulp_push_disk_images.py (hunk 11 above)
Comment (happybhati) [reply scripts/python/tasks/internal/pulp_push_disk_images.py]: Done, thanks

File: scripts/python/tasks/internal/pulp_push_disk_images.py (hunk 12 above)
Comment (happybhati) [reply scripts/python/tasks/internal/pulp_push_disk_images.py]: Switched to `push_artifacts._check_cert_expiration` via a small `_validate_certificates` wrapper for the three mounts. Thanks for the pointer!

File: scripts/python/helpers/oras_utils.py (hunk 9 above)
Comment (johnbieren) [reply scripts/python/helpers/oras_utils.py:83]: thanks, the try finally tricked me

**Review by johnbieren — DISMISSED** [review]

**Review by twaugh — APPROVED** [review]

**Comment by davidmogar** [comment]
## Production Approval Record

| Field | Value |
|---|---|
| **Action** | APPROVED |
| **Reviewer** | @davidmogar |
| **Timestamp** | 2026-06-24T14:23:16.673Z |

**Comment by happybhati** [comment]
Rebased onto `main` and resolved conflicts in `oras_utils.py` and `test_oras_utils.py` (merged `oras_resolve` from main with our `oras_pull`). Force-push dismissed approvals — @swickersh @johnbieren could you please re-approve? Thanks!

**Review by swickersh — APPROVED** [review]
Is the plan to hold off on the catalog referencing this until we have e2e for this pipeline? 
Or do you have a way to mock up a full disk image release?
These questions don't affect this PR being merged... just the catalog counterpart.

**Comment by johnbieren** [comment]
For next time, I think if you have to rebase in a way that resets approvals, you should get them all again before merging (vs just 1). I know it is a bit of a gray area

**Comment by happybhati** [comment]
> For next time, I think if you have to rebase in a way that resets approvals, you should get them all again before merging (vs just 1). I know it is a bit of a gray area

Ah, my bad, thanks for the feedback. I will definitely keep it in mind next time.
